pycolmap/custom_bundle_adjustment.py

```diff
@@ -0,0 +1,440 @@
+        ba_options_tmp.solver_options.max_linear_solver_iterations = 200
+
+    # Avoid degeneracies in bundle adjustment
+    reconstruction.filter_observations_with_negative_depth()
+
+    # Configure bundle adjustment
+    ba_config = pycolmap.BundleAdjustmentConfig()
+    for image_id in reg_image_ids:
+        ba_config.add_image(image_id)
+
+    # Fix the existing images, if option specified
+    if mapper_options.fix_existing_images:
+        for image_id in reg_image_ids:
+            if image_id in mapper.existing_image_ids:
+                ba_config.set_constant_cam_pose(image_id)
+
+    # Fix 7-DOFs of the bundle adjustment problem
+    ba_config.set_constant_cam_pose(reg_image_ids[0])
+    if (not mapper_options.fix_existing_images) or (
+        reg_image_ids[1] not in mapper.existing_image_ids
+    ):
+        ba_config.set_constant_cam_positions(reg_image_ids[1], [0])
+
+    # Run bundle adjustment
+    summary = solve_bundle_adjustment(reconstruction, ba_options_tmp, ba_config)
+    logging.info("Global Bundle Adjustment")
+    logging.info(summary.BriefReport())
+
+
+def iterative_global_refinement(
+    mapper,
+    max_num_refinements,
+    max_refinement_change,
+    mapper_options,
+    ba_options,
+    tri_options,
+    normalize_reconstruction=True,
+):
+    """Equivalent to mapper.iterative_global_refinement(...)"""
+    reconstruction = mapper.get_reconstruction()
+    mapper.complete_and_merge_tracks(tri_options)
+    logging.verbose(
+        1,
+        "=> Retriangulated observations: {0}".format(
+            mapper.retriangulate(tri_options)
+        ),
+    )
+    for i in range(max_num_refinements):
+        num_observations = reconstruction.compute_num_observations()
+        # mapper.adjust_global_bundle(mapper_options, ba_options)
+        adjust_global_bundle(mapper, mapper_options, ba_options)
+        if normalize_reconstruction:
+            reconstruction.normalize()
+        num_changed_observations = mapper.complete_and_merge_tracks(tri_options)
+        num_changed_observations += mapper.filter_points(mapper_options)
+        changed = (
+            0
+            if num_observations == 0
+            else num_changed_observations / num_observations
+        )
+        logging.verbose(1, "=> Changed observations: {0:.6f}".format(changed))
+        if changed < max_refinement_change:
+            break
+
+
+def adjust_local_bundle(
+    mapper, mapper_options, ba_options, tri_options, image_id, point3D_ids
+):
+    """Equivalent to mapper.adjust_local_bundle(...)"""
+    reconstruction = mapper.get_reconstruction()
+    assert reconstruction is not None
+    report = pycolmap.LocalBundleAdjustmentReport()
+
+    # Find images that have most 3D points with given image in common
+    local_bundle = mapper.find_local_bundle(mapper_options, image_id)
+
+    # Do the bundle adjustment only if there is any connected images
+    if len(local_bundle) > 0:
```

Comment:
May I ask if there are any reasons for such changes (also the following similar changes). Doesnt it reduce the readability? Or are there some rules of convention for such syntax?

Reply:
Google's style guide: https://google.github.io/styleguide/pyguide.html#2144-decision

Reply:
Sure. This will be updated. 

However, in the guide it says "When handling integers, implicit false may involve more risk than benefit (i.e., accidentally handling None as 0). You may compare a value which is known to be an integer (and is not the result of len()) against the integer 0." And also the case in the example suggests the same? So should we still update the following changes? e.g. "if report.num_adjusted_observations > 0:" to "if report.num_adjusted_observations:"

Reply:
This has been resolved.